libs/validate_results.py

```diff
@@ -0,0 +1,112 @@
+
+"""
+A set of functions to validate the datasets we prepare
+"""
+
+OUTPUT_HEADERS_SHARE = [
+    'Province/State',
+    "Country/Region",
+    "Last Update",
+    "Latitude",
+    "Longitude",
+    "State/County FIPS Code",
+    'State Intervention',
+    '16d-HSPTLZD',
+    '32d-HSPTLZD',
+    '16d-LACKBEDS',
+    '32d-LACKBEDS',
+    'MEAN-HOSP',
+    'MEAN-DEATHS',
+    'PEAK-HOSP',
+    'PEAK-DEATHS',
+    "Current Deaths",
+    "Current Confirmed",
+    "Current Recovered",
+    "Current Active",
+    "Combined Key",
+    "County"
+]
+
+
+OUTPUT_HEADERS_STATES = OUTPUT_HEADERS_SHARE + []
+OUTPUT_HEADERS_COUNTIES = OUTPUT_HEADERS_SHARE + []
+
+EXPECTED_MISSING_STATES = set([
+    'Northern Mariana Islands', 'American Samoa', 'Virgin Islands', 'Puerto Rico', 'Guam'
+])
+
+EXPECTED_MISSING_STATES_FROM_COUNTES = set([
+    'District of Columbia'
+])
+
+class DataExportException(Exception): 
+    def __init__(self, key, message):
+        self.key = key
+        self.message = message
+
+def __raise_error_if_incorrect_headers(key, headers_list, df): 
+    extra_df_columns = set(df.columns) - set(headers_list)
+    missing_df_columns = set(headers_list) - set(df.columns)
+    if extra_df_columns or missing_df_columns: 
+        raise DataExportException(key, f"Exported headers don't match expected headers. Didn't have headers {missing_df_columns}, had extra columns {extra_df_columns}")
+
+def __raise_error_if_not_data_from_all_states(key, df, expected_missing): 
+    states = set(us_state_abbrev.keys())
+    states_in_df = set(df['Province/State'].unique())
+
+    missing_states_in_df = states - states_in_df - expected_missing
+    extra_states_in_df = states_in_df - states
+    if missing_states_in_df or extra_states_in_df: 
+        raise DataExportException(key, f"Missing Data from states: {missing_states_in_df}. Have extra states in df {extra_states_in_df}")
+
+def validate_states_df(key, states_df):
+    # assert the headers are what we expect
+    __raise_error_if_incorrect_headers(key, OUTPUT_HEADERS_STATES, states_df)
+  
+    # assert there is data from each of the states
+    __raise_error_if_not_data_from_all_states(key, states_df, EXPECTED_MISSING_STATES)
+    
+    # assert no duplicated states
+    if len(states_df['Province/State'].unique()) != len(states_df['Province/State']): 
+        raise DataExportException(key, f"Duplicated State Data: {states_df['Province/State'] - states_df['Province/State'].unique()}")
+
+def validate_counties_df(key, counties_df):
+    # assert the headers are what we expect
+    __raise_error_if_incorrect_headers(key, OUTPUT_HEADERS_COUNTIES, counties_df)
+
+    # assert data from each of the states
+    __raise_error_if_not_data_from_all_states(key, counties_df, EXPECTED_MISSING_STATES.union(EXPECTED_MISSING_STATES_FROM_COUNTES))
+
+    # assert no duplicated counties
+    if len(counties_df['State/County FIPS Code'].unique()) != len(counties_df['State/County FIPS Code']): 
```

Comment:
just for reference:
https://github.com/covid-projections/covid-data-model/blob/master/libs/datasets/dataset_utils.py#L93
this does a similar thing here

Reply:
cool i think the duplication is fine here